firebase-firestore/src/main/java/com/google/firebase/firestore/local/SQLiteRemoteDocumentCache.java

```diff
@@ -136,20 +143,38 @@ public MaybeDocument get(DocumentKey documentKey) {
                 return;
               }
 
-              MaybeDocument maybeDoc = decodeMaybeDocument(row.getBlob(1));
-              if (!(maybeDoc instanceof Document)) {
-                return;
-              }
-
-              Document doc = (Document) maybeDoc;
-              if (!query.matches(doc)) {
-                return;
-              }
-
-              results.put(doc.getKey(), doc);
+              byte[] rawContents = row.getBlob(1);
```

Comment:
This one method is now huge, tightly intermingling threading implementation details with the query processing. Most of this could be generic. There are other circumstances where we potentially process large numbers of rows (e.g. the mutation queue) and this could be extended to those places. Probably even more importantly, if we pull this out into something separate we could test it directly :-).

One thing that's preventing this from being generic is the intermediate results are stored in a concurrent map, which ties the results to the query structure. However, we don't need this to be a map of key to document until we're assembling the result. We could just as easily publish to each document into a dequeue.

Additional notes:
  * This implementation leaves time on the table by not starting query matching until all the protos are decoded. The Firestore worker thread that was processing query results can't start assembling the result before the worker threads are done.
  * Also, we're performing query matching serially on the Firestore worker, when we could be performing it in the parallel block.
  * This is a very low number of threads that are all doing significant work per item. We could probably get away with just a simple lock wrapping our output without building an intermediate `ConcurrentHashMap`.

If we were targeting Java 8, I'd suggest that you've essentially reimplemented a subset of `java.util.stream`. There are ports of that to Android API < 24, but that's a lot of code for something that's pretty small.

At a high level this could be phrased as map/reduce, where the map phase is parsing and filtering (first by `instanceof Document`, second by `query.matches`). The reduce phase is the assembly of the `matchingDocuments` map.

I suggest a few refactoring experiments:

  * See if pulling query matching into the worker threads helps
  * See if directly assembling the `ImmutableSortedMap` while holding a lock works
      * If not, see if a regular `Deque` of matching docs while holding a lock is OK
      * Alternatively try a `ConcurrentLinkedQueue`
  * Then try to extract a component that does the parallel processing in a way that's more self contained

WDYT?

Reply:
I'll try to make this more generic, but I didn't want this to turn into a 500 line change with multiple new classes. I'll give it another go though.

Re: Query processing. Query matching is not idempotent (because of https://github.com/firebase/firebase-android-sdk/blob/745272a1e939448ca2f92934af2aa73ffb8c427b/firebase-firestore/src/main/java/com/google/firebase/firestore/core/Query.java#L54) and If we do want to perform this concurrently, then I would rather tackle this in a follow-up PR.

Reply:
Does this turn into a problem in practice? What if we access the orderBy in this thread before enqueuing any work?

Reply:
@mikelehen mostly convince me that Query matching is not a problem in practice. I also believe that Query matching itself is not a huge bottleneck, but it doesn't hurt to also execute it in the background, especially given that we are already running a task there.

I rewrote the code in this function to use the Tasks API. By that I mean I added APIs to Util that mirror their equivalent methods of the Task API, but with some significant differences:

- My version of `await` can be called on the main thread to support unit testing of the RemoteDocumentStore.
- My version of `whenAllComplete` doesn't execute the continuation on the main queue, which blocks our SpecTests on Robolectric.

If you like this approach, I can write dedicated tests for these helpers. We might also want to move them to a dedicated place (such as TaskUtil).